pageserver/src/tenant/timeline.rs

```diff
@@ -4375,6 +4375,9 @@ enum OpenLayerAction {
     None,
 }
 
+/// Order of updates in a batch are forced using this structure
+pub type OrderedBatchUpdates = BTreeMap<(Lsn, Key), Value>;
```

Comment:
BTreeMap is quite a bit more expensive than Vec in terms of memory allocations -- did you find a case where the existing Vec was sometimes out of order?

The intent of the existing code is that updates are implicitly ordered by LSN because we only insert while traversing the WAL in LSN order.

Reply:
> BTreeMap is quite a bit more expensive than Vec in terms of memory allocations -- did you find a case where the existing Vec was sometimes out of order?
No, I haven't, this addresses the issue https://github.com/neondatabase/neon/issues/6707.

I agree that constructing BTreeMap is expensive, maybe replacing the existing `pending_updates` as HashMap with BTreeMap, that would avoid sorting afterward and would also force an order by LSN?

Reply:
Thanks for linking the issue.

I think the gist of that issue was that maybe we would have a wrapper that asserts ordering at insert time (i.e. the value appended is higher-LSN than the last value in the vector) -- but @VladLazar has more context.  Vlad - can you take the review of this PR?

Reply:
John is correct. The intention is to assert if the Lsns are out of order. Let's use a `VecMap<Lsn, (Key, Value)>` (see `libs/utils/src/vec_map.rs`). You'll have to implement [FromIterator](https://doc.rust-lang.org/std/iter/trait.FromIterator.html) for `VecMap` to get `collect` to work.